otp4gb/cost.py

```diff
@@ -0,0 +1,358 @@
+        Factors and other parameters for calculating the
+        generalised cost.
+    aggregation_method : AggregationMethod
+        Aggregation method used for generalised cost matrix.
+    workers : int, default 0
+        Number of threads to create during calculations.
+    """
+    LOG.info("Calculating costs for %s", matrix_file.name)
+
+    lock = threading.Lock()
+    response_file = matrix_file.with_name(matrix_file.name + "-response_data.jsonl")
+    with open(response_file, "wt", encoding=util.TEXT_ENCODING) as responses:
+        iterator = util.multithread_function(
+            workers,
+            calculate_costs,
+            jobs,
+            dict(
+                response_file=responses,
+                lock=lock,
+                generalised_cost_parameters=generalised_cost_parameters.copy(),
+            ),
+        )
+
+        matrix_data = []
+        for res in tqdm.tqdm(
+            iterator,
+            total=len(jobs),
+            desc="Calculating costs",
+            dynamic_ncols=True,
+            smoothing=0,
+        ):
+            matrix_data.append(res)
+
+    LOG.info("Written responses to %s", response_file)
+    _write_matrix_files(matrix_data, matrix_file, aggregation_method)
+
+
+def iterate_responses(response_file: io.TextIOWrapper) -> Iterator[CostResults]:
+    """Iterate through and parse responses JSON lines file.
+
+    Parameters
+    ----------
+    response_file : io.TextIOWrapper
+        JSON lines file to read from.
+
+    Yields
+    ------
+    CostResults
+        Cost results for a single OD pair.
+    """
+    for line in response_file:
+        yield CostResults.parse_raw(line)
+
+
+def cost_matrix_from_responses(
+    responses_files: list[pathlib.Path],
+    matrix_file: pathlib.Path,
+    aggregation_method: AggregationMethod,
+) -> None:
+    """Create cost matrix CSV from responses JSON lines file.
+
+    Parameters
+    ----------
+    responses_files : list[pathlib.Path]
+        List of Path(s) to JSON lines file containing `CostResults`.
+    matrix_file : pathlib.Path
+        Path to CSV file to output cost metrics to.
+    aggregation_method : AggregationMethod
+        Aggregation method used for generalised cost matrix.
+    """
+
+    matrix_data = []
+
+    LOG.info("Compiling %s OTP response files", len(responses_files))
+    for responses_file in responses_files:
+        LOG.info("Processing file %s",
+                 responses_file)
+        with open(responses_file, "rt", encoding=util.TEXT_ENCODING) as responses:
+            for line in tqdm.tqdm(
+                    responses, desc="Calculating cost matrix", dynamic_ncols=True
+            ):
+                results = CostResults.parse_raw(line)
+                # TODO(MB) Recalculate generalised cost if new parameters are provided
+                matrix_data.append(_matrix_costs(results))
+
+        _write_matrix_files(matrix_data, matrix_file, aggregation_method)
```

Comment:
[a7c4a04](https://github.com/Transport-for-the-North/otp4gb-py/commit/a7c4a04af989cf43eefd0df140b90cc52e10bdee)

Great spot, I meant to have placed this outside of the for loop for this reason and missed this, so this would have been an issue for a while!